### PR TITLE
[shopsys] nominal promo code discount is limited by applicable products

### PR DESCRIPTION
**Pull request by AndrejBlaho**
| Q             | A
| ------------- | ---
|Description, reason for the PR| When using nominal promo code, the total discount amount was not correctly limited by total value of applicable products.
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://ab-nominal-discount-limited-to-applicable-products.odin.shopsys.cloud
  - https://cz.ab-nominal-discount-limited-to-applicable-products.odin.shopsys.cloud

### CODE REVIEW

File: UPGRADE-15.0.md [hunk 1]
```diff
@@ -1058,6 +1058,10 @@ Follow the instructions in relevant sections, e.g. `shopsys/coding-standards` or
 
 -   see #project-base-diff to update your project
 
+#### nominal promo code discount is limited by applicable products ([#3227](https://github.com/shopsys/shopsys/pull/3227))
+
+-   see #project-base-diff to update your project
```
Comment (grossmannmartin) [inline UPGRADE-15.0.md]: no change was made in the projet-base package, so I believe the upgrade notes are not necessary

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 2]
```diff
@@ -80,20 +82,32 @@ public function createAndAddOrderItemData(
...
         $locale = $domainConfig->getLocale();
         $domainId = $domainConfig->getId();
 
         $defaultVat = $this->vatFacade->getDefaultVatForDomain($domainId);
 
+        $totalApplicableProductsPrice = $this->calculateTotalApplicableProductsPrice($orderData, $validProductIds);
+
+        $discountValue = Money::create($promoCodeLimit->getDiscount());
+
+        if ($discountValue > $totalApplicableProductsPrice) {
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Money objects cannot be compared like this. 

if you try this 
```
$money = Money::create(1000);
$money2 = Money::createFromFloat(1000.00, 2);
d($money < $money2);
```

it prints `true` although the value should be the same

proper comparison is 
```
if ($discountValue->isLessThan($totalApplicableProductsPrice)) {
```

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 3]
```diff
@@ -116,4 +130,22 @@ protected function createDiscountOrderItemData(
 
         return $discountOrderItemData;
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\OrderData $orderData
+     * @param array $validProductIds
+     * @return \Shopsys\FrameworkBundle\Component\Money\Money
+     */
+    protected function calculateTotalApplicableProductsPrice(OrderData $orderData, array $validProductIds): Money
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: The current name suggests the method returns the `Price` object. 
What about the name `calculateTotalApplicableProductsPriceAmountWithVat`?

It's long, but the intention is clearer, I think.

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 4]
```diff
@@ -116,4 +130,22 @@ protected function createDiscountOrderItemData(
...
+    {
+        $totalPrice = Money::zero();
+
+        foreach ($orderData->items as $item) {
+            if (in_array($item->product->getId(), $validProductIds, true)) {
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: in theory, the $items property may contain different item types than products (for example when someone changes the order of middlewares), and the product property may be null by definition. 

I suggest using 

```
foreach ($orderData->getItemsByType(OrderItemTypeEnum::TYPE_PRODUCT) as $item) {
    if (in_array($item->product?->getId(), $validProductIds, true)) {

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 5]
```diff
@@ -116,4 +130,22 @@ protected function createDiscountOrderItemData(
...
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\OrderData $orderData
+     * @param array $validProductIds
+     * @return \Shopsys\FrameworkBundle\Component\Money\Money
+     */
+    protected function calculateTotalApplicableProductsPrice(OrderData $orderData, array $validProductIds): Money
+    {
+        $totalPrice = Money::zero();
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: maybe `$totalPriceAmountWithVat`?

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 6]
```diff
@@ -116,4 +130,22 @@ protected function createDiscountOrderItemData(
 
         return $discountOrderItemData;
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Order\OrderData $orderData
+     * @param array $validProductIds
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]:
```suggestion
     * @param int[] $validProductIds
```

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 7]
```diff
@@ -80,20 +82,32 @@ public function createAndAddOrderItemData(
...
      * @return \Shopsys\FrameworkBundle\Model\Order\Item\OrderItemData|null
      */
     protected function createDiscountOrderItemData(
         PromoCode $promoCode,
         PromoCodeLimit $promoCodeLimit,
         DomainConfig $domainConfig,
+        OrderData $orderData,
+        array $validProductIds,
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: What about calculating the totalApplicableProductsPrice in the `createAndAddOrderItemData()` method and passing only the calculated value? What do you think?

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: UPGRADE-15.0.md (hunk 1 above)
Comment (AndrejBlaho) [reply UPGRADE-15.0.md]: Ah. I'm sorry, now it is removed.

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 2 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Ah i see, I've fixed it.

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 3 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Sure, i've changed it.

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 4 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Understood and changed.

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 5 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: sure

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 6 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Fixed.

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 7 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Sure, that seems cleaner.

**Review by grossmannmartin — APPROVED** [review]

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 8]
```diff
@@ -62,10 +62,13 @@ public function createAndAddOrderItemData(
         PromoCodeLimit $promoCodeLimit,
         OrderProcessingData $orderProcessingData,
     ): void {
+        $totalApplicableProductsPrice = $this->calculateTotalApplicableProductsPriceAmountWithVat($orderData, $validProductIds);
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]:
```suggestion
        $totalApplicableProductsPriceAmountWithVat = $this->calculateTotalApplicableProductsPriceAmountWithVat($orderData, $validProductIds);
```

same in the `createDiscountOrderItemData()` method definition 😇

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php [hunk 9]
```diff
@@ -80,20 +83,28 @@ public function createAndAddOrderItemData(
...
         $locale = $domainConfig->getLocale();
         $domainId = $domainConfig->getId();
 
         $defaultVat = $this->vatFacade->getDefaultVatForDomain($domainId);
 
+        $discountValue = Money::create($promoCodeLimit->getDiscount());
+
+        if ($discountValue->isLessThan($totalApplicableProductsPrice)) {
+            $discountValue = $totalApplicableProductsPrice;
+        }
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: 🤔  I think we've missed the logic flaw here... 
Shouldn't it be `$discount->isGreaterThan($totalApplicableProductsPrice)`? 

Now I have an order completely free for the 500€ nominal coupon 😆 

Snímek obrazovky 2024-06-19 v 9 13 20

**Review by grossmannmartin — CHANGES_REQUESTED** [review]
can you please also check the standards?

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 8 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Ah, the hard part of programmers life - naming variables correctly. I hope it is ok now. 😁

File: packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php (hunk 9 above)
Comment (AndrejBlaho) [reply packages/framework/src/Model/Order/Processing/OrderProcessorMiddleware/ApplyNominalPromoCodeMiddleware.php]: Indeed we did. Now it should be ok.

**Review by grossmannmartin — APPROVED** [review]

**Pull request merged**